File: 309.最佳买卖股票时机含冷冻期.go
/*
 * @lc app=leetcode.cn id=309 lang=golang
 *
 * [309] 最佳买卖股票时机含冷冻期
 */

// @lc code=start
// 代录；动态规划；
func maxProfit(prices []int) int {
	dp := make([][]int, len(prices))
	for i := range dp {
		dp[i] = make([]int, 4)
	}
	dp[0][0] = -prices[0]
	// dp[0][1] = 0 // 默认值为0；
	// dp[0][2] = 0
	// dp[0][3] = 0
	for i := 1; i < len(prices); i++ {
		dp[i][0] = max(dp[i-1][0], max(dp[i-1][3]-prices[i], dp[i-1][1]-prices[i]))
		dp[i][1] = max(dp[i-1][1], dp[i-1][3])
		dp[i][2] = dp[i-1][0] + prices[i]
		dp[i][3] = dp[i-1][2]
	}
	return max(dp[len(prices)-1][1], max(dp[len(prices)-1][2], dp[len(prices)-1][3]))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

File: leetcode_recordes/309.最佳买卖股票时机含冷冻期.py
#
# @lc app=leetcode.cn id=309 lang=python3
#
# [309] 最佳买卖股票时机含冷冻期
#


# @lc code=start


# 未完成；
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        k = n // 2  # 最多交易次数

        dp = [0] * (1 + 2 * k + 1)  # 最后位标记是否是今天卖出的；初始为0
        # 初始化
        for i in range(1 + 2 * k, 2):
            dp[i] = -prices[0]

        for i in range(1, n):
            for j in range(1, 1 + 2 * k):
                if dp[-1] == 1:
                    dp[-1] = 0  # 度过冷冻期；把买入的跳过，但卖出的还可以；
                    dp[j] = max(dp[j], dp[j - 1] + prices[i])
                    continue
                if j % 2:
                    dp[j] = max(dp[j], dp[j - 1] - prices[i])
                else:
                    dp[j] = max(dp[j], dp[j - 1] + prices[i])
                    if dp[j] < dp[j - 1] + prices[i]:
                        dp[-1] = 1  # 标记卖出；

        return dp[-2]


# @lc code=end
